Source/stores.h

```diff
@@ -12,96 +12,140 @@
 #include "control.h"
 #include "engine.h"
 #include "engine/clx_sprite.hpp"
+#include "engine/random.hpp"
+#include "options.h"
+#include "qol/stash.h"
+#include "towners.h"
 #include "utils/attributes.h"
 
 namespace devilution {
 
-#define WITCH_ITEMS 25
-#define SMITH_ITEMS 25
-#define SMITH_PREMIUM_ITEMS 15
-#define STORE_LINES 104
+/** @brief Number of player items that display in stores (Inventory slots and belt slots) */
+const int NumPlayerItems = (NUM_XY_SLOTS - (SLOTXY_EQUIPPED_LAST + 1));
+
+constexpr int NumSmithBasicItems = 19;
+constexpr int NumSmithBasicItemsHf = 24;
+
+constexpr int NumSmithItems = 6;
+constexpr int NumSmithItemsHf = 15;
+
+constexpr int NumHealerItems = 17;
+constexpr int NumHealerItemsHf = 19;
+constexpr int NumHealerPinnedItems = 2;
+constexpr int NumHealerPinnedItemsMp = 3;
+
+constexpr int NumWitchItems = 17;
+constexpr int NumWitchItemsHf = 24;
+constexpr int NumWitchPinnedItems = 3;
+
+constexpr int NumBoyItems = 1;
+
+constexpr int NumStoreLines = 104;
+
+extern _talker_id TownerId;
 
 enum class TalkID : uint8_t {
-	None,
-	Smith,
-	SmithBuy,
-	SmithSell,
-	SmithRepair,
-	Witch,
-	WitchBuy,
-	WitchSell,
-	WitchRecharge,
+	Exit,
+	MainMenu,
+	BasicBuy,
+	Buy,
+	Sell,
+	Repair,
+	Recharge,
+	Identify,
+	IdentifyShow,
+	Stash,
 	NoMoney,
 	NoRoom,
 	Confirm,
-	Boy,
-	BoyBuy,
-	Healer,
-	Storyteller,
-	HealerBuy,
-	StorytellerIdentify,
-	SmithPremiumBuy,
 	Gossip,
-	StorytellerIdentifyShow,
-	Tavern,
-	Drunk,
-	Barmaid,
+	Invalid,
 };
 
-/** Currently active store */
-extern TalkID ActiveStore;
+enum class ItemLocation {
+	Inventory,
+	Belt,
+	Body
+};
 
-/** Current index into PlayerItemIndexes/PlayerItems */
-extern DVL_API_FOR_TEST int CurrentItemIndex;
-/** Map of inventory items being presented in the store */
-extern int8_t PlayerItemIndexes[48];
-/** Copies of the players items as presented in the store */
-extern DVL_API_FOR_TEST Item PlayerItems[48];
+struct StoreMenuOption {
+	TalkID action;
+	std::string text;
+};
 
-/** Items sold by Griswold */
-extern Item SmithItems[SMITH_ITEMS];
-/** Number of premium items for sale by Griswold */
-extern int PremiumItemCount;
-/** Base level of current premium items sold by Griswold */
-extern int PremiumItemLevel;
-/** Premium items sold by Griswold */
-extern Item PremiumItems[SMITH_PREMIUM_ITEMS];
+struct TownerLine {
+	const std::string menuHeader;
+	const StoreMenuOption *menuOptions;
+	size_t numOptions;
+};
 
-/** Items sold by Pepin */
-extern Item HealerItems[20];
+struct IndexedItem {
+	Item *itemPtr;         // Pointer to the original item
+	ItemLocation location; // Location in the player's inventory (Inventory, Belt, or Body)
+	int index;             // Index in the corresponding array

```

Comment:
Please use doc blocks instead of inline comments so that the documentation is picked up properly by tools.